### PR TITLE
Jepsen setup for Chardonnay

### PR DESCRIPTION
**Pull request by purujit**
I have refactored the docker file to use a separate base layer for jepsen which is still built off of Debian bookworm but has jepsen specific dependencies and configuration installed. I am using a `BUILD_TYPE` arg to isolate the jepsen specific changes.
Also fixed a start up bug in epoch publisher that was preventing this job from coming up if epoch service wasn't already running.

### CODE REVIEW

**Comment by purujit** [comment]
@tamer-eldeeb @yanniszark Got it working. All the servers come up and epoch gets published. I'll work on polishing this up. If you have ideas around how to structure the Dockerfile so that we can still share the common parts, let me know - we probably should not load all of the jepsen deps in the main image but I guess it is not harmful to leave it in there either.

File: setup-jepsen.sh
Comment (purujit) [inline setup-jepsen.sh:1]: This file is copied from the jepsen repo.

**Comment by purujit** [comment]
@tamer-eldeeb  I have found a decent solution to isolate the jepsen related changes in the Dockerfile. PTAL when your tests are done. I can demo it to the group next meeting.

File: chardonnay.service [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[Unit]
```
Comment (yanniszark) [inline chardonnay.service:1]: Why do we need a systemd service for chardonnay?

**Review by yanniszark — COMMENTED** [review]
Hey @purujit thanks a lot and sorry for the wait on the review.
One thing I don't understand 100% is the reason for refactoring the images to use systemd, as well as the purpose of the systemd chardonnay service in every image.
Generally, systemd does similar things to k8s/docker (e.g., health / readiness checks), so imo we should only introduce the extra complexity of dealing with systemd (which in this case seems to be maintaining an extra image for compatibility + logging I would guess) only if necessary (e.g., we might need to start a lot of different binaries / services in the same container).

File: setup-jepsen.sh
Comment (purujit) [reply setup-jepsen.sh:1]: Left over from debugging. Fixed. now.

**Comment by purujit** [comment]
> Hey @purujit thanks a lot and sorry for the wait on the review. One thing I don't understand 100% is the reason for refactoring the images to use systemd, as well as the purpose of the systemd chardonnay service in every image. Generally, systemd does similar things to k8s/docker (e.g., health / readiness checks), so imo we should only introduce the extra complexity of dealing with systemd (which in this case seems to be maintaining an extra image for compatibility + logging I would guess) only if necessary (e.g., we might need to start a lot of different binaries / services in the same container).

More details in the other comments but I mostly tried to make minimal changes to the Docker setup in the jepsen repo and systemd is not strictly necessary - it was just one of the options I saw people use in other repos. Open to other ideas - we can always iterate.

File: Dockerfile [hunk 2]
```diff
@@ -40,63 +41,125 @@ COPY . .
 # Build the project
 RUN cargo build --release
 
+###############################################################################
+# node-jepsen #################################################################
+###############################################################################
+# This is a arm64 build of jgoerzen/debian-base-minimal:bookworm.
+FROM purujit/chardonnay:debian-base-minimal AS debian-addons
```
[the remark this answers is not in the document]
Comment (yanniszark) [reply Dockerfile:49]: That sounds good to me! I would say maybe leave a comment there for future reference.

**Comment by yanniszark** [comment]
> without interrupting the boot-debian-base script that seems to be core to the base image for jepsen
Very nice, makes sense! I wasn't sure if boot-debian-base was something Jepsen needed. I'll read up on it a bit later, but maybe it's worth leaving a small comment about it and why Jepsen needs it / what's the general deployment arch for it.

I'll take a closer look and test later today since I'll be flying soon.

**Comment by yanniszark** [comment]
Took a final pass, looks good to me! AFAIK, boot-debian-base isn't really a Jepsen thing, it just sets up systemd. The core issue here as far as i understand is that Jepsen requires ssh access. To satisfy that we need to also run sshd in our containers, which means that now we're running 2 processes instead of 1. Thus, we introduce systemd to take care of the process management. Kinda unfortunate that Jepsen can't use kubectl :/ 

Nice idea for the Dockerfile structuring. The format is quite limiting, so this is a neat solution.

**Review by yanniszark — APPROVED** [review]
lgtm

**Pull request merged**